utils/src/Vulkan/Utils/Reflect.hs

```diff
@@ -0,0 +1,414 @@
+  let pipelineShaderStageCreateInfos = (SomeStruct <$>) . makePipelineShaderStageCreateInfos stage entryPoints
+  ShaderInfo {..}
+
+makeDescriptorInfo :: Vector (Shader, Reflection) -> Vector (DescriptorSetLayoutCreateInfo '[])
+makeDescriptorInfo = makeDescriptorSetLayoutCreateInfos . join . V.map (makeDescriptorSetLayoutBindings . (stage :: Shader -> ShaderStage) . fst <*> fromMaybe [] . ubos . snd <*> fromMaybe [] . textures . snd)
+
+makeInputInfo :: Vector (Shader, Reflection) -> Maybe (SomeStruct PipelineVertexInputStateCreateInfo)
+makeInputInfo = (SomeStruct <$>) . makePipelineVertexInputStateCreateInfo . join . V.fromList . catMaybes . (inputs . snd <$>) . filter ((== Vert) . (stage :: Shader -> ShaderStage) . fst) . V.toList
+
+makeShaderModuleCreateInfo :: "code" ::: B.ByteString -> ShaderModuleCreateInfo '[]
+makeShaderModuleCreateInfo code = zero { code = code }
+
+makePipelineShaderStageCreateInfo :: ShaderStage -> ShaderModule -> EntryPoint -> PipelineShaderStageCreateInfo '[]
+makePipelineShaderStageCreateInfo stage shaderModule EntryPoint {..} = zero
+  { stage = convertStage stage
+  , module' = shaderModule
+  , name = B.pack . T.unpack $ name
+  }
+
+makePipelineShaderStageCreateInfos :: ShaderStage -> Vector EntryPoint -> ShaderModule -> Vector (PipelineShaderStageCreateInfo '[])
+makePipelineShaderStageCreateInfos stage entryPoints shaderModule = V.map (makePipelineShaderStageCreateInfo stage shaderModule) entryPoints
+
+-- https://github.com/KhronosGroup/GLSL/blob/master/extensions/khr/GL_KHR_vulkan_glsl.txt
+-- https://vulkan.lunarg.com/doc/view/1.2.135.0/linux/tutorial/html/08-init_pipeline_layout.html
+makeUboDescriptorSetLayoutBinding :: ShaderStage -> Ubo -> (Int, DescriptorSetLayoutBinding)
+makeUboDescriptorSetLayoutBinding stage Ubo {..} = (set, zero
+  { binding = fromIntegral binding
+  , descriptorType = DESCRIPTOR_TYPE_UNIFORM_BUFFER
+  , descriptorCount = maybe 1 (V.sum . (fromIntegral <$>)) array
+  , stageFlags = convertStage stage
+  })
+
+data TextureDescriptorType
+  = Sampler2D
+
+instance IsString TextureDescriptorType where
+  fromString = \case
+    "sampler2D" -> Sampler2D
+    unsupport -> error $ "TextureDescriptorType not support '" <> unsupport <> "'"
+
+convertTextureDescriptorType :: TextureDescriptorType -> DescriptorType
+convertTextureDescriptorType = \case
+  Sampler2D -> DESCRIPTOR_TYPE_COMBINED_IMAGE_SAMPLER
+
+makeTextureDescriptorSetLayoutBinding :: ShaderStage -> Texture -> (Int, DescriptorSetLayoutBinding)
+makeTextureDescriptorSetLayoutBinding stage Texture {..} = (set, zero
+  { binding = fromIntegral binding
+  , descriptorType = convertTextureDescriptorType . fromString . T.unpack $ type'
+  , descriptorCount = maybe 1 (V.sum . (fromIntegral <$>)) array
+  , stageFlags = convertStage stage
+  })
+
+makeDescriptorSetLayoutBindings :: ShaderStage -> Vector Ubo -> Vector Texture -> Vector (Int, DescriptorSetLayoutBinding)
+makeDescriptorSetLayoutBindings stage ubos textures = do
```

Comment:
Using `do` with pure functions is confusing. You can put away helpers under a `where` block and leave only the function body as the definition.

Reply:
fixed. though I am not sure is this `let-in-do for no monad function` usage a bad practice.

I prefer to use `let` for value binding, and use `where` for function binding. And in `do block` it can prevent write a `in` after `let`

Reply:
and both `let` and `do notation` makes me feel like I am doing some imperative programming